xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -430,15 +526,226 @@ private void handleConfigNotFound(StatusException exception) {
+          }
+        }
+        SslContextProviderSupplier defaultSupplier =
+                selector.getDefaultSslContextProviderSupplier();
+        if (defaultSupplier != null) {
+          toRelease.add(defaultSupplier);
+        }
+      }
+      return toRelease;
+    }
+
+    private void releaseSuppliersInFlight() {
+      SslContextProviderSupplier supplier;
+      for (FilterChain filterChain : filterChains.values()) {
+        supplier = filterChain.getSslContextProviderSupplier();
+        if (supplier != null) {
+          supplier.close();
+        }
+      }
+      if (defaultFilterChain != null
+              && (supplier = defaultFilterChain.getSslContextProviderSupplier()) != null) {
+        supplier.close();
+      }
+    }
+
+    private final class RouteDiscoveryState implements RdsResourceWatcher {
+      private final String resourceName;
+      private final FilterChain filterChain;
+      private final List<NamedFilterConfig> httpFilterConfigs;
+
+      private RouteDiscoveryState(String resourceName, FilterChain filterChain,
+          List<NamedFilterConfig> httpFilterConfigs) {
+        this.resourceName = checkNotNull(resourceName, "resourceName");
+        this.filterChain = checkNotNull(filterChain, "filterChain");
+        this.httpFilterConfigs = checkNotNull(httpFilterConfigs, "httpFilterConfigs");
+      }
+
+      @Override
+      public void onChanged(final RdsUpdate update) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            if (!routeDiscoveryStates.containsKey(resourceName)) {
+              return;
+            }
+            ServerRoutingConfig routingConfig =
+                    ServerRoutingConfig.create(httpFilterConfigs, update.virtualHosts);
+            handleConfigDiscovered(filterChain, routingConfig);
+          }
+        });
+      }
+
+      @Override
+      public void onResourceDoesNotExist(final String resourceName) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            if (!routeDiscoveryStates.containsKey(resourceName)) {
+              return;
+            }
+            StatusException statusException = Status.UNAVAILABLE.withDescription(
+                    "Rds " + resourceName + " unavailable").asException();
+            handleConfigNotFound(statusException);
+          }
+        });
+      }
+
+      @Override
+      public void onError(final Status error) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            if (!routeDiscoveryStates.containsKey(resourceName)) {
+              return;
+            }
+            boolean isPermanentError = isPermanentError(error);
+            logger.log(Level.FINE, "{0} error from XdsClient: {1}",
+                    new Object[]{isPermanentError ? "Permanent" : "Transient", error});
+            if (isPermanentError) {
```

Comment:
I don't understand how this logic came about. I'd expect the same behavior independent of the status code.

Reply:
It inherits from LDS error handling, which inherits from the old behaviour before xds server refactoring.
It looks watcher `onError`  is only triggered StreamObserver of the xdsClient's own stream onError.  And xdsClient would always retry. And our LB watchers treat this as the well known transient_ failure. So it may make sense to handle `onError` more lightly here as well as a transient error, e.g. log and notify and do nothing?

Reply:
I think log+fail RPCs for this filter chain with a canned message would be appropriate. I'd say that'd be similar to `onResourceDoesNotExist()`, but it seems `onResourceDoesNotExist()` hasn't been updated yet based on the decision to only fail RPCs that select the broken filter chain.

Reply:
oh i missed this decision. I'll try to update it then.
We are only taking about RDS for this partial failure handling, right? 
Is it appropriate to kill all RPCs for LDS `onResourceNotExist` and shutdown delegate server?, and also how was permanent error handling on LDS `onError`?

Reply:
Yes, it only impacts RDS. LDS behaves as before and can shut down the port.